shared-bindings/aes/aes.c

```diff
@@ -0,0 +1,355 @@
+  aes_obj_t *self = MP_OBJ_TO_PTR(aes_obj);
+
+  int mode = mp_obj_get_int(mode_obj);
+  switch (mode) {
+  case AES_MODE_CBC: /* CBC */
+    break;
+  case AES_MODE_ECB: /* ECB */
+    break;
+  case AES_MODE_CTR: /* CTR */
+    break;
+  default:
+    mp_raise_TypeError(translate("Requested AES mode is unsupported"));
+  }
+
+  common_hal_aes_set_mode(self, mode);
+  return mp_const_none;
+}
+
+MP_DEFINE_CONST_FUN_OBJ_2(aes_set_mode_obj, aes_set_mode);
+
+
+STATIC byte *duplicate_data(const byte *src_buf, size_t len) {
+  byte *dest_buf = m_new(byte, len);
+  memcpy(dest_buf, src_buf, len);
+  return dest_buf;
+}
+
+//|   .. method:: encrypt_in_place(buf)
+//|
+//|      Encrypt the provided buffer in-place without copying it first.
+//|      The buffer must be a multiple of 16 bytes.
+//|
+STATIC mp_obj_t aes_encrypt_in_place(mp_obj_t aes_obj, mp_obj_t buf) {
+  if (!MP_OBJ_IS_TYPE(aes_obj, &aes_aes_type)) {
+    mp_raise_TypeError_varg(translate("Expected a %q"), aes_aes_type.name);
+  }
+  // Convert parameters into expected types.
+  aes_obj_t *aes = MP_OBJ_TO_PTR(aes_obj);
+  mp_buffer_info_t bufinfo;
+  mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_READ);
+  if ((bufinfo.len & (AES_BLOCKLEN - 1)) != 0) {
+    mp_raise_ValueError(translate("Buffer must be a multiple of 16 bytes"));
+  }
+
+  common_hal_aes_encrypt(aes, (uint8_t *)bufinfo.buf, bufinfo.len);
+  return mp_const_none;
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(aes_encrypt_in_place_obj,
+                                 aes_encrypt_in_place);
+
+//|   .. method:: encrypt_into(src, dest)
+//|
+//|      Encrypt the buffer from ``src`` into ``dest``.
+//|      The buffers must be a multiple of 16 bytes, and must
+//|      be equal length
+//|
+STATIC mp_obj_t aes_encrypt_into(mp_obj_t aes_obj, mp_obj_t src,
+                                 mp_obj_t dest) {
+  if (!MP_OBJ_IS_TYPE(aes_obj, &aes_aes_type)) {
+    mp_raise_TypeError_varg(translate("Expected a %q"), aes_aes_type.name);
+  }
+  // Convert parameters into expected types.
+  aes_obj_t *aes = MP_OBJ_TO_PTR(aes_obj);
+
+  mp_buffer_info_t srcbufinfo;
+  mp_get_buffer_raise(src, &srcbufinfo, MP_BUFFER_READ);
+  if ((srcbufinfo.len & (AES_BLOCKLEN - 1)) != 0) {
+    mp_raise_ValueError(translate("Buffer must be a multiple of 16 bytes"));
+  }
+
+  mp_buffer_info_t destbufinfo;
+  mp_get_buffer_raise(dest, &destbufinfo, MP_BUFFER_READ);
+  if (destbufinfo.len != srcbufinfo.len) {
+    mp_raise_ValueError(
+        translate("Source and destination buffers must be the same length"));
+  }
+
+  memcpy(destbufinfo.buf, srcbufinfo.buf, srcbufinfo.len);
+
+  common_hal_aes_decrypt(aes, (uint8_t *)destbufinfo.buf, destbufinfo.len);
+  return mp_const_none;
+}
+
+STATIC MP_DEFINE_CONST_FUN_OBJ_3(aes_encrypt_into_obj, aes_encrypt_into);
+
+//|   .. method:: encrypt(buf)
+//|
+//|      Encrypt the buffer and return a copy of the encrypted data.
+//|      The buffer must be a multiple of 16 bytes.
+//|
```

Comment:
I wouldn't even provide this. Folks should manage the buffers themselves.

Reply:
This comes from when we were going for `pycrypto` compatibility.  Since that's no longer a goal, I'll remove it.